core/Network/TLS/Handshake/Client.hs

```diff
@@ -880,20 +873,17 @@ handshakeClient13' cparams ctx groupSent usedCipher usedHash = do
         usingHState ctx getGroupPrivate >>= fromServerKeyShare serverKeyShare
 
     makeEarlySecret = do
-        secret <- usingHState ctx getTLS13Secret
-        case secret of
-          EarlySecret sec -> do
-              mSelectedIdentity <- usingState_ ctx getTLS13PreSharedKey
-              case mSelectedIdentity of
-                Nothing                          ->
-                    return (hkdfExtract usedHash zero zero, False)
-                Just (PreSharedKeyServerHello 0) -> do
-                    unless (B.length sec == hashSize) $
-                        throwCore $ Error_Protocol ("selected cipher is incompatible with selected PSK", True, IllegalParameter)
-                    usingHState ctx $ setTLS13HandshakeMode PreSharedKey
-                    return (sec, True)
-                Just _                           -> throwCore $ Error_Protocol ("selected identity out of range", True, IllegalParameter)
-          _ -> return (hkdfExtract usedHash zero zero, False)
+        earlySecretPSK@(BaseSecret sec) <- usingHState ctx getTLS13EarlySecret
+        mSelectedIdentity <- usingState_ ctx getTLS13PreSharedKey
+        case mSelectedIdentity of
+          Nothing                          ->
+              return (calcEarlySecret choice Nothing, False)
+          Just (PreSharedKeyServerHello 0) -> do
+              unless (B.length sec == hashSize) $
+                  throwCore $ Error_Protocol ("selected cipher is incompatible with selected PSK", True, IllegalParameter)
+              usingHState ctx $ setTLS13HandshakeMode PreSharedKey
+              return (earlySecretPSK, True)
+          Just _                           -> throwCore $ Error_Protocol ("selected identity out of range", True, IllegalParameter)
 
```

Comment:
You are right. And done.

Reply:
I don't understand this completely. But did you suggest the following logic?

```haskell
    makeEarlySecret = do
         mEarlySecretPSK <- usingHState ctx getTLS13EarlySecret
         case mEarlySecretPSK of
           Nothing -> return (initEarlySecret choice Nothing, False)
           Just earlySecretPSK@(BaseSecret sec) -> do
               mSelectedIdentity <- usingState_ ctx getTLS13PreSharedKey
               case mSelectedIdentity of
                 Nothing                          ->
                     return (initEarlySecret choice Nothing, False)
                 Just (PreSharedKeyServerHello 0) -> do
                     unless (B.length sec == hashSize) $
                         throwCore $ Error_Protocol ("selected cipher is incompatible with selected PSK", True, IllegalParameter)
                     usingHState ctx $ setTLS13HandshakeMode PreSharedKey
                     return (earlySecretPSK, True)
                 Just _                           -> throwCore $ Error_Protocol ("selected identity out of range", True, IllegalParameter)
```

Reply:
Yes exactly. Earlier in the code `setTLS13EarlySecret` is called conditionally based on `pskInfo` and this decision can be kept here too. Otherwise it changes the behaviour faced to a misbehaving server.

Reply:
Replaced.